java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -279,6 +279,8 @@ private void inferReturnTypeFromInferedArgs(MethodInvocationTree tree, Resolve.E
         MethodJavaType methodType = (MethodJavaType) resolution.type();
         if(!methodType.resultType.isTagged(JavaType.DEFERRED)) {
           registerType(tree, resolve.applySubstitution(methodType.resultType, typeSubstitution));
+          // change type associated to identifier as it may have been inferred in the process
+          registerType(identifier, resolution.type());
```

Comment:
Am I correct in saying that you ended up associating methodType to the identifier of the method in order to access it from check ? That is far from obvious and deserve at least a comment (I am even wondering about a fixme here, this sounds a bit like a hack to shove the methodType into the semantic API)